In [1]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [2]:
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader

from models.unet_3d import UNet3D
from utils.unets_helper_functions import (
    set_seed,
    train_one_epoch,
    validate_one_epoch,
    save_checkpoint,
    PatchDataset,
    evaluate_full_volume
)


In [3]:
set_seed(42)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)

Using device: cuda


In [4]:
FOLD = 0

with open(f"../data/splits/fold_{FOLD}/train.txt") as f:
    train_cases = f.read().splitlines()

with open(f"../data/splits/fold_{FOLD}/val.txt") as f:
    val_cases = f.read().splitlines()

print("Train cases:", len(train_cases))
print("Val cases:", len(val_cases))

Train cases: 33
Val cases: 9


In [5]:
PATCH_SIZE = 80
PATCHES_PER_CASE = 6

train_dataset = PatchDataset(train_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = PATCHES_PER_CASE,
                        patch_size = PATCH_SIZE,
                        augment=False)

val_dataset = PatchDataset(val_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = 1,
                        patch_size = PATCH_SIZE,
                        augment=False)


train_loader = DataLoader(
    train_dataset,
    batch_size=2,
    shuffle=True,
    num_workers=2,      
    pin_memory=True    
)

val_loader = DataLoader(
    val_dataset,
    batch_size=1,
    num_workers=2,    
    pin_memory=True
)

In [6]:
model = UNet3D(
    in_channels=1,
    out_channels=7,
    base_filters=16
).to(device)

print("Model initialized.")

Model initialized.


In [7]:
optimizer = optim.Adam(model.parameters(), lr=1e-4)
weights = torch.tensor([0.1, 1, 1, 1, 1, 1, 1]).to(device)
ce_loss = nn.CrossEntropyLoss(weight=weights)
scaler = torch.cuda.amp.GradScaler()

C:\Users\dhanu\AppData\Local\Temp\ipykernel_6232\1548250201.py:4: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [8]:
EPOCHS = 40

PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "unet_fold0")

os.makedirs(SAVE_DIR, exist_ok=True)
best_val_loss = float("inf")

for epoch in range(EPOCHS):

    train_loss = train_one_epoch(
        model,
        train_loader,
        optimizer,
        scaler,
        ce_loss,
        device
    )

    val_loss, val_dice = validate_one_epoch(
        model,
        val_loader,
        ce_loss,
        device
    )

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")

    # Save best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "best_model.pth")
        )
        print("Saved Best Model")

  0%|          | 0/99 [00:00<?, ?it/s]

100%|██████████| 99/99 [08:11<00:00,  4.97s/it]



Epoch 0
Train Loss: 2.9405
Val Loss: 2.7977
Per Class Dice: [6.16968465e-02 2.24897451e-06 1.23847832e-03 5.67575315e-03
 1.03658654e-02 1.02101282e-01]
Saved Best Model


100%|██████████| 99/99 [05:41<00:00,  3.45s/it]



Epoch 1
Train Loss: 2.6727
Val Loss: 2.6591
Per Class Dice: [0.04467258 0.11677807 0.00324895 0.05416658 0.00413119 0.13718109]
Saved Best Model


100%|██████████| 99/99 [06:38<00:00,  4.03s/it]



Epoch 2
Train Loss: 2.5366
Val Loss: 2.4165
Per Class Dice: [0.11628464 0.28593742 0.0166852  0.21530491 0.0748069  0.19932798]
Saved Best Model


100%|██████████| 99/99 [07:04<00:00,  4.29s/it]



Epoch 3
Train Loss: 2.4211
Val Loss: 2.3856
Per Class Dice: [0.08234054 0.16374994 0.0330828  0.15827343 0.06849436 0.24619789]
Saved Best Model


100%|██████████| 99/99 [07:44<00:00,  4.70s/it]



Epoch 4
Train Loss: 2.2861
Val Loss: 2.3860
Per Class Dice: [0.15896708 0.27654273 0.01765811 0.10941072 0.00090446 0.19571382]


100%|██████████| 99/99 [07:22<00:00,  4.47s/it]



Epoch 5
Train Loss: 2.1906
Val Loss: 2.1625
Per Class Dice: [0.30516007 0.47255224 0.17524627 0.40544854 0.15942028 0.34574738]
Saved Best Model


100%|██████████| 99/99 [06:49<00:00,  4.13s/it]



Epoch 6
Train Loss: 2.1224
Val Loss: 2.0107
Per Class Dice: [0.41976452 0.53671544 0.06055489 0.21600872 0.28524111 0.50410448]
Saved Best Model


100%|██████████| 99/99 [07:05<00:00,  4.30s/it]



Epoch 7
Train Loss: 2.0572
Val Loss: 1.9826
Per Class Dice: [0.39907846 0.54103447 0.02537141 0.44467256 0.22095375 0.4856571 ]
Saved Best Model


100%|██████████| 99/99 [06:11<00:00,  3.76s/it]



Epoch 8
Train Loss: 1.9738
Val Loss: 2.0162
Per Class Dice: [0.33099409 0.46724779 0.33333351 0.08349131 0.26391547 0.28752173]


100%|██████████| 99/99 [05:55<00:00,  3.59s/it]



Epoch 9
Train Loss: 1.8852
Val Loss: 1.8821
Per Class Dice: [0.29466489 0.76877477 0.55555559 0.18862205 0.19183301 0.31233915]
Saved Best Model


100%|██████████| 99/99 [05:50<00:00,  3.54s/it]



Epoch 10
Train Loss: 1.8379
Val Loss: 1.8561
Per Class Dice: [0.29969883 0.82572193 0.55555593 0.28031006 0.16501515 0.32198928]
Saved Best Model


100%|██████████| 99/99 [05:14<00:00,  3.17s/it]



Epoch 11
Train Loss: 1.7857
Val Loss: 1.8576
Per Class Dice: [0.38033046 0.92751224 0.3333335  0.11635929 0.12505681 0.16333322]


100%|██████████| 99/99 [06:17<00:00,  3.82s/it]



Epoch 12
Train Loss: 1.7068
Val Loss: 1.7695
Per Class Dice: [0.17072402 0.91751856 0.66666676 0.81024237 0.17774333 0.17023094]
Saved Best Model


100%|██████████| 99/99 [05:52<00:00,  3.57s/it]



Epoch 13
Train Loss: 1.6736
Val Loss: 1.6428
Per Class Dice: [0.59776333 0.78850918 0.55555572 0.55983381 0.34301909 0.33233971]
Saved Best Model


100%|██████████| 99/99 [05:44<00:00,  3.48s/it]



Epoch 14
Train Loss: 1.6105
Val Loss: 1.6461
Per Class Dice: [0.47209194 0.82326674 0.55555556 0.38257893 0.2625997  0.3279235 ]


100%|██████████| 99/99 [06:51<00:00,  4.16s/it]



Epoch 15
Train Loss: 1.5684
Val Loss: 1.5801
Per Class Dice: [0.44068764 0.89786326 0.55555556 0.328711   0.36397731 0.3337118 ]
Saved Best Model


 91%|█████████ | 90/99 [06:01<00:36,  4.01s/it]


KeyboardInterrupt: 

In [8]:
#  RESUME TRAINING 
PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "unet_fold0")

start_epoch = 0
best_val_loss = float("inf")

resume_path = os.path.join(SAVE_DIR, "best_model.pth")  

if os.path.exists(resume_path):
    print("Loading checkpoint:", resume_path)

    checkpoint = torch.load(resume_path, map_location=device)

    model.load_state_dict(checkpoint["model_state_dict"])
    optimizer.load_state_dict(checkpoint["optimizer_state_dict"])

    start_epoch = checkpoint["epoch"] + 1
    best_val_loss = checkpoint["best_val_loss"]

    print(f"Resuming from epoch {start_epoch}")
else:
    print("No checkpoint found, starting fresh.")


Loading checkpoint: c:\Users\dhanu\Desktop\mini-project-1\experiments\unet_fold0\best_model.pth
Resuming from epoch 16


C:\Users\dhanu\AppData\Local\Temp\ipykernel_6232\3872517435.py:13: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(resume_path, map_location=device)


In [10]:
EPOCHS = 40
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "unet_fold0")

os.makedirs(SAVE_DIR, exist_ok=True)

for epoch in range(EPOCHS):

    train_loss = train_one_epoch(
        model,
        train_loader,
        optimizer,
        scaler,
        ce_loss,
        device
    )

    val_loss, val_dice = validate_one_epoch(
        model,
        val_loader,
        ce_loss,
        device
    )

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")

    # Save best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "best_model.pth")
        )
        print("Saved Best Model")

100%|██████████| 99/99 [05:04<00:00,  3.07s/it]



Epoch 0
Train Loss: 1.5484
Val Loss: 1.5046
Per Class Dice: [0.68387318 0.83034942 0.66666695 0.6676505  0.26061589 0.6391875 ]
Saved Best Model


100%|██████████| 99/99 [04:39<00:00,  2.82s/it]



Epoch 1
Train Loss: 1.4880
Val Loss: 1.4771
Per Class Dice: [0.61233602 0.73231394 0.66666667 0.62847349 0.39454936 0.52752268]
Saved Best Model


100%|██████████| 99/99 [05:31<00:00,  3.35s/it]



Epoch 2
Train Loss: 1.4527
Val Loss: 1.4139
Per Class Dice: [0.6461296  0.76155548 0.33333337 0.49962241 0.54034884 0.49641883]
Saved Best Model


100%|██████████| 99/99 [05:24<00:00,  3.28s/it]



Epoch 3
Train Loss: 1.4282
Val Loss: 1.4190
Per Class Dice: [0.48990955 0.5562625  0.55555556 0.60516604 0.3777376  0.62308905]


100%|██████████| 99/99 [05:10<00:00,  3.14s/it]



Epoch 4
Train Loss: 1.3826
Val Loss: 1.4128
Per Class Dice: [0.49291444 0.886401   0.77777778 0.47838701 0.44630146 0.47006216]
Saved Best Model


100%|██████████| 99/99 [05:17<00:00,  3.21s/it]



Epoch 5
Train Loss: 1.3490
Val Loss: 1.3412
Per Class Dice: [0.5374416  0.88854282 0.44444448 0.78922346 0.60701043 0.55903174]
Saved Best Model


100%|██████████| 99/99 [05:23<00:00,  3.27s/it]



Epoch 6
Train Loss: 1.3256
Val Loss: 1.2851
Per Class Dice: [0.71313597 0.79816214 0.22222223 0.84197533 0.72656165 0.73768354]
Saved Best Model


100%|██████████| 99/99 [05:35<00:00,  3.39s/it]



Epoch 7
Train Loss: 1.3070
Val Loss: 1.2806
Per Class Dice: [0.72932251 0.77116408 0.11111117 0.63732788 0.66899742 0.72298757]
Saved Best Model


100%|██████████| 99/99 [05:26<00:00,  3.30s/it]



Epoch 8
Train Loss: 1.2766
Val Loss: 1.2987
Per Class Dice: [0.54710702 0.7163334  0.55555556 0.61137455 0.49124374 0.65034805]


100%|██████████| 99/99 [04:56<00:00,  2.99s/it]



Epoch 9
Train Loss: 1.2510
Val Loss: 1.2400
Per Class Dice: [0.79767447 0.81232416 0.66666667 0.77462394 0.78325086 0.83056052]
Saved Best Model


100%|██████████| 99/99 [04:59<00:00,  3.02s/it]



Epoch 10
Train Loss: 1.2303
Val Loss: 1.2374
Per Class Dice: [0.69337036 0.70846235 0.66666667 0.76650771 0.76329765 0.81160707]
Saved Best Model


100%|██████████| 99/99 [05:51<00:00,  3.55s/it]



Epoch 11
Train Loss: 1.2148
Val Loss: 1.2453
Per Class Dice: [0.40831025 0.94564205 0.66666667 0.460665   0.47777541 0.52203105]


100%|██████████| 99/99 [05:58<00:00,  3.62s/it]



Epoch 12
Train Loss: 1.1765
Val Loss: 1.2076
Per Class Dice: [0.85313822 0.93757183 0.77777778 0.82849836 0.85968385 0.84776556]
Saved Best Model


100%|██████████| 99/99 [3:10:06<00:00, 115.22s/it]    



Epoch 13
Train Loss: 1.1733
Val Loss: 1.1529
Per Class Dice: [0.61652825 0.79263424 0.77777778 0.80420431 0.80297275 0.67783851]
Saved Best Model


100%|██████████| 99/99 [05:25<00:00,  3.29s/it]



Epoch 14
Train Loss: 1.1454
Val Loss: 1.1695
Per Class Dice: [0.51162857 0.94019857 0.55555556 0.75140805 0.49669741 0.46204764]


100%|██████████| 99/99 [05:46<00:00,  3.50s/it]



Epoch 15
Train Loss: 1.1213
Val Loss: 1.1325
Per Class Dice: [0.57965542 0.90678732 0.55555556 0.57636189 0.37869175 0.34226895]
Saved Best Model


100%|██████████| 99/99 [05:07<00:00,  3.11s/it]



Epoch 16
Train Loss: 1.1249
Val Loss: 1.1728
Per Class Dice: [0.50770809 0.9320397  0.77777778 0.43991039 0.47790684 0.51249377]


100%|██████████| 99/99 [05:32<00:00,  3.36s/it]



Epoch 17
Train Loss: 1.1126
Val Loss: 1.0835
Per Class Dice: [0.71039515 0.80943689 0.44444446 0.74623034 0.68130711 0.73208634]
Saved Best Model


100%|██████████| 99/99 [05:39<00:00,  3.43s/it]



Epoch 18
Train Loss: 1.0873
Val Loss: 1.1011
Per Class Dice: [0.55385218 0.86105768 0.44444446 0.67850044 0.57747435 0.69982294]


100%|██████████| 99/99 [06:02<00:00,  3.66s/it]



Epoch 19
Train Loss: 1.0745
Val Loss: 1.0723
Per Class Dice: [0.79860601 0.66649068 0.66666668 0.74471747 0.82346847 0.79963046]
Saved Best Model


100%|██████████| 99/99 [05:09<00:00,  3.13s/it]



Epoch 20
Train Loss: 1.0596
Val Loss: 1.1095
Per Class Dice: [0.60360608 0.92512898 0.55555556 0.73302861 0.73686269 0.60006187]


100%|██████████| 99/99 [06:08<00:00,  3.72s/it]



Epoch 21
Train Loss: 1.0373
Val Loss: 1.0058
Per Class Dice: [0.8837536  0.89933846 0.44444452 0.82913163 0.73001726 0.82373444]
Saved Best Model


100%|██████████| 99/99 [05:41<00:00,  3.45s/it]



Epoch 22
Train Loss: 1.0362
Val Loss: 1.0840
Per Class Dice: [0.621275   0.91434297 0.66666667 0.85546311 0.66364276 0.58844902]


100%|██████████| 99/99 [06:30<00:00,  3.94s/it]



Epoch 23
Train Loss: 1.0142
Val Loss: 0.9739
Per Class Dice: [0.93196557 0.89578203 0.77777778 0.81762567 0.79093319 0.78349146]
Saved Best Model


100%|██████████| 99/99 [05:37<00:00,  3.40s/it]



Epoch 24
Train Loss: 1.0147
Val Loss: 0.9549
Per Class Dice: [0.79493628 0.85621588 0.55555556 0.72787144 0.76251099 0.87643252]
Saved Best Model


100%|██████████| 99/99 [05:39<00:00,  3.43s/it]



Epoch 25
Train Loss: 0.9910
Val Loss: 0.9900
Per Class Dice: [0.4698914  0.93698092 0.66666667 0.62718891 0.61738763 0.4481755 ]


 58%|█████▊    | 57/99 [03:20<02:27,  3.51s/it]


KeyboardInterrupt: 

In [7]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = UNet3D().to(device)

model_path = "../experiments/unet_fold0/best_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


Model loaded for testing.


C:\Users\dhanu\AppData\Local\Temp\ipykernel_26616\3740780336.py:6: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


In [12]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.8768281592122439), np.float64(0.8437210088186547), np.float64(3.752345201679755e-09), np.float64(0.6803811831187557), np.float64(0.7632063586670482), np.float64(0.8866197826492611)]
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.8541621387051287), np.float64(0.7910046020743209), np.float64(3.4340659222731253e-09), np.float64(0.7966350783982997), np.float64(0.7892194255090138), np.float64(0.8584967265779521)]
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8112759478637509), np.float64(0.7793083657775087), np.float64(3.409478338187936e-09), np.float64(0.7484893400134025), np.float64(0.8342920508460101), np.float64(0.894142520799234)]
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_14 Dice: [np.float64(0.8581276144882672), np.float64(0.7282570974330137), np.float64(4.496402

In [8]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Volume: 272x357x357 | Patches: 1521
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.8846944948857364), np.float64(0.8300474889696018), np.float64(3.752345201679755e-09), np.float64(0.7200824221587554), np.float64(0.7860180481422901), np.float64(0.8857292362307195)]
Volume: 243x383x383 | Patches: 1568
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.8613414334177978), np.float64(0.77783259350614), np.float64(3.4340659222731253e-09), np.float64(0.7976704489418263), np.float64(0.8100541281589908), np.float64(0.8742818126010822)]
Volume: 264x333x333 | Patches: 1296
Unique pred labels: [0 1 2 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.8380634391117197), np.float64(0.8159921580899089), np.float64(3.409478338187936e-09), np.float64(0.7476562062832447), np.float64(0.8480318884988028), np.float64(0.9127361113456184)]
Volume: 248x395x395 | Patches: 1800
Unique pred labels: [0 

## UNet 3D Patch-Based Validation Performance

The model was trained for 40 epochs using patch-based validation.  
Below are the final patch validation metrics obtained at the best validation loss epoch.

### Final Training Statistics (Epoch 39 - Best Model)

- Train Loss: 1.0147
- Val Loss: 0.9549

## Observations

- Validation loss steadily improved toward final epochs.
- Model converged well overall.
- Some class instability observed during training.


Train Loss 0.9783 | Val Loss 0.9308

# Evaluating on Full Volume
~ (patch = 80 , stride = 60)

| Organ ID | Organ (Based on Your Map) | Mean Dice   | Std Dev |
| -------- | ------------------------- | ----------- | ------- |
| 1        | Bone Mandible             | **0.8685**  | 0.0306  |
| 2        | Brainstem                 | **0.8169**  | 0.0250  |
| 3        | Spinal Cord               | **~0.0000** | ~0.0000 |
| 4        | Parotid Left              | **0.7135**  | 0.0452  |
| 5        | Parotid Right             | **0.7690**  | 0.0764  |
| 6        | Oral Cavity               | **0.8376**  | 0.0531  |

SpinalCord Fails maybe beacuse It is extremely thin structure
Very small voxel count
Hard to capture with 80³ patches
May not be sufficiently sampled